README
NAME
    OpenResty - General-purpose web service platform for web applications

VERSION
    This document describes OpenResty 0.5.12 released on Nov 20, 2009.

DESCRIPTION
    This module implements the server-side OpenResty web service protocol.
    It provides scriptable and extensible web services for both server-side
    and client-side (pure AJAX) web applications.

    Currently this module can serve as a public web interface to a
    distributed or desktop PostgreSQL database system. In particular, it
    provides roles, models, views, actions, captchas, the minisql language,
    and many more to the web users.

    "Another framework?" No, no, no, not all!

    OpenResty is *not* a web application framework like Jifty or Catalyst.
    Rather, it is

    *   A REST wrapper for relational databases

    *   A web runtime for 100% JavaScript web sites and other RIAs.

    *   A "meta web site" supporting other sites via web services.

    *   A handy personal or company database which can be accessed from
        anywhere on the web.

    *   A (sort of) competitor for the Facebook Data Store API.

    We're already running an instance of the OpenResty server on our Yahoo!
    China's production machines:

    <http://api.openresty.org/=/version>

    And there're several (pure-client-side) web sites alreadying taking
    advantage of the services:

    OpenResty's admin site
        <http://openresty.org/admin/>

    agentzh's blog and EEEE Works' blog
        <http://blog.agentzh.org>

        <http://eeeeworks.org>

    Yisou BBS
        <http://www.yisou.com/opi/post.html>

    See OpenResty::Spec::Overview for more detailed information.

    OpenResty::CheatSheet also provides a good enough summary for the REST
    interface.

    You'll find my slides for the D2 conference interesting as well:

    <http://agentzh.org/misc/openresty-d2.pdf>

    or the original XUL version:

    <http://agentzh.org/misc/openresty-d2/openresty-d2.xul> (Firefox
    required)

    Another good introduction to OpenResty's REST API is summerized in the
    slides for my Y!ES talk and my Beijing Perl Workshop 2008 talk:

    <http://agentzh.org/misc/openresty-yes.pdf>

    and a more pretty (XUL) version can be got from here:

    <http://agentzh.org/misc/openresty-yes/openresty-yes.xul> (Firefox
    required)

    There're also a few interesting discussions about OpenResty on my blog
    site:

    "OpenResty versus Google App Engine"
        <http://blog.agentzh.org/#post-75>

    "Google's crawlers captured OpenResty's API!"
        <http://blog.agentzh.org/#post-79>

    "Video for my D2 talk about OpenResty and its nifty apps"
        <http://blog.agentzh.org/#post-81>

    "The first yahoo.cn feature that is powered by OpenResty"
        <http://blog.agentzh.org/#post-86>

    "Client-side web site DIY" (Chinese)
        <http://blog.agentzh.org/#post-80>

    "OpenResty 平台相关资料" (Chinese)
        <http://www.eeeeworks.org/#post-6>

CAVEATS
    This library is still in the beta phase and the API is still in flux.
    We're just following the "release early, releaes often" guideline. So
    please check back often ;)

INSTALLATION
    Please see OpenResty::Spec::Install for details :)

SOURCE TREE STRUCTURE
    bin/
        contains some command-line utilities, among which the openresty is
        the most important one.

    lib/
        contains all the server code, mostly Perl.

    haskell/
        contains the RestyScript compiler for OpenResty written in Haskell.
        Support for both OpenResty Views and Actions is provided.

        See haskell/README for more details.

    font/
        contains the font file (*.ttf) for captcha generation.

    etc/
        contains the config files, openresty.conf and site_openresty.conf.
        The latter one takes precedence over the former.

    grammar/
        contains Parse::Yapp grammar files for the old OpenResty View (or
        minisql) compiler.

    t/  contains the test suite.

    demo/
        contains a bunch of OpenResty demo apps.

    inc/
        generated by Module::Install for CPAN building system.

PERFORMANCE
    OpenResty takes runtime performance very seriously because we have to
    run it on our not-so-good servers and support lots of Yahoo! China's
    online products with very heavy traffic.

    OpenResty prefers modules with XS over pure Perl ones and uses cache
    aggressively. It's also in favor of source-filter based solutions
    provided by Filter::QuasiQuote to reduce the length of subroutine
    calling chains and the number of indirections. Finally, the restyscript
    compiler is also written in carefully optimized Haskell code to maximize
    speed.

    The benchmark results for OpenResty 0.5.3's test suite on a PentiumIV
    3.0GHz machine is given below:

    in-process frontend + PgMocked backend
            DELETE: 4 ms (157 trials)
            POST: 23 ms (493 trials)
            PUT: 5 ms (132 trials)
            GET: 4 ms (648 trials)

    lighttpd fastcgi frontend + local Pg backend
            DELETE: 29 ms (193 trials)
            POST: 30 ms (815 trials)
            PUT: 11 ms (138 trials)
            GET: 9 ms (763 trials)

    lighttpd fastcgi frontend + remote PgFarm backend
            DELETE: 99 ms (193 trials)
            POST: 98 ms (815 trials)
            PUT: 41 ms (138 trials)
            GET: 24 ms (763 trials)

SOURCE CONTROL
    For the very latest version of this module, check out the source from
    the Git repos below:

    <http://github.com/agentzh/openresty/tree/master>

    There is anonymous access to all. If you'd like a commit bit, please let
    us know. :)

Mailing list
    Subscribe to the "openresty" Google Group here:

      L<http://groups.google.com/group/openresty>

Project Roadmap
    Below is a list of currently planned release milestones (but it's also
    supposed to change as we go):

    0.5.x (Where we are)
        Action API and an enhanced version of the Model API.

    0.6.x
        Migrate the View handler to the same style and implementation of the
        Action handler, i.e., using explicit parameter list and taking
        advantage of the Haskell version of the restyscript compiler.

        Compiling view definition to native PostgreSQL functions is also
        supposed to realize in this series.

    0.7.x
        Attachment API, which supports binary file uploading and
        downloading.

    0.8.x
        Mail API, which introduces builtin Models for email sentbox and
        inbox based on third-party POP3/STMP servers.

        It will also allow actions to be triggered and/or confirmed by
        emails.

    0.9.x
        Prophet/Git integration.

    Please don't hesitate to tell us what you love to see in future releases
    of OpenResty ;)

TODO
    For the project's TODO list, please check out
    <http://svn.openfoundry.org/openapi/trunk/TODO>

BUGS
    There must be some serious bugs lurking somewhere given the current
    status of the implementation and test suite.

    Please report bugs or send wish-list to
    <http://rt.cpan.org/NoAuth/Bugs.html?Dist=OpenResty>.

AUTHORS
    Agent Zhang (agentzh) "<agentzh at yahoo dot cn>"
    Xunxin Wan (万珣新) "<wanxunxin at gmail dot com >"
    chaoslawful (王晓哲) "<chaoslawful at gmail dot com>"
    Lei Yonghua (leiyh)
    Laser Henry (laser) "<laserhenry at gmail dot com>"
    Yu Ting (yuting) "<yuting at yahoo dot cn>"

    For a complete list of the contributors, please see
    <http://svn.openfoundry.org/openapi/trunk/AUTHORS>.

License and Copyright
    OpenResty is licensed under the BSD License:

    Copyright (c) 2007-2008, Yahoo! China EEEE Works, Alibaba Inc. All
    rights reserved.

    Copyright (c) 2007-2008, Agent Zhang (agentzh). All rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are
    met:

    *   Redistributions of source code must retain the above copyright
        notice, this list of conditions and the following disclaimer.

    *   Redistributions in binary form must reproduce the above copyright
        notice, this list of conditions and the following disclaimer in the
        documentation and/or other materials provided with the distribution.

    *   Neither the name of the Yahoo! China EEEE Works, Alibaba Inc. nor
        the names of its contributors may be used to endorse or promote
        products derived from this software without specific prior written
        permission.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
    IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
    TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
    PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER
    OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
    EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
    PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
    PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
    LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
    NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

SEE ALSO
    OpenResty::Spec::Overview, openresty, OpenResty::Spec::REST_cn,
    OpenResty::CheatSheet, WWW::OpenResty, WWW::OpenResty::Simple.



Changes
Change Log for OpenResty

0.5.12 Nov 20 2009
* removed purge from openresyt.js since it is obsolete and useless.
* applied the patch from kindy++ for removing the dojo dependency from openresty.js.
* no global deflate filter in the apache config because it may cause problems.
* applied the POD patch from Frank Wiegand.

0.5.11 May 21 2009
* emphasized that OpenResty requires the line "AllowEncodedSlashes On" in the Apache2 config file.
* now we croak loudly while failing to load handler classes.

0.5.10 May 21 2009
* added a sample custom handler class OpenResty::Handler::Shell.
* added the "openresty" mailing list page to the POD.

0.5.9 May 6 2009
* updated CheatSheet.pod to reflect the latest OpenResty API.

0.5.8 May 6 2009
* added support for frontend.unlimited config directive, which lists the OpenResty accounts without ordinary limits. (xunxin)
* now we use dist_dir rather than module_dir in shared stuff locating. (agentzh)

0.5.7 May 4 2009
* added missing files to MANIFEST.

0.5.6 May 3 2009
* added docs OpenResty::Spec::Install::Apache and OpenResty::Spec::Install::Lighttd.

0.5.5 May 1 2009
* updated the source repos to the new GitHub address.
* fixed the encoding name in GettingStarted_cn.pod.

0.5.4 May 1 2009
* now we honor the FastCGI protocol and send the Status header to the fastcgi manager. [agentzh]
* updated the RSS links in the blog demo sites. [agentzh]
* now we drop model in the cascaded fashion. [agentzh]
* added applied_to column to _access table so that we can add ACL rules containing ip address validations. [xunxin]
* replaced JSON::XS with JSON::Syck module in import-model.pl to allow loose JSON syntax in input data files. [xunxin]
* implemented captcah+password login method and the test file t/55-login-by-captchapassword.t for it. [xunxin]
* now we can add OpenResty account by --wrap-legacy option to reuse legacy Pg schemas insteading creating new ones. [xunxin]
* now we support no limit accounts. [xunxin]
* minor performance tuning by eliminating some unnecessary DB queries. [xunxin]
* replaced Cache::Memcached::Fast by Cache::Memcached::libmemcached because the former might miss existing keys when access rate is very high. [xunxin]
* minor UI fix in the Admin site for IE. [xunxin]
* preliminary support for compiled actions. [xunxin]

0.5.3 Nov 30 2008
* switched the license from GPLv2 to BSD.
* updated demo/Blog/script/init.pl to use the new column default value syntax.
* added the benchmark results for various configurations to the main POD.
* added the milestones and release plans to the main POD.
* mentioned the #openresty IRC channel in the main POD.

0.5.2 Nov 27 2008
* added support for the "unique" attribute of Model columns and updated the Admin site for it.
* rewrote the "default" attribute to support expressions and updated the Admin site for it.
* Bumped the version number of metamodel to 0.008 and we now no longer make use of the meta tables _models and _columns.
* now the -f option of the bin/openresty script supports absolute path and relative path of config files. No etc/ prefix assumed now.
* the proprietary YLogin API is mostly working.
* modify the default value of ACCOUNT_CMD_COUNT_LIMIT to 20.
* demo/Onccf - minor css fixes.
* added --- status section to t::OpenResty so as to test status code.
* skipped clkao++'s sanity check in http mode.

0.5.1 Nov 21 2008
* added support for builtin vars in the main url, as in
    GET /=/feed/Post/_user/agentzh.Public
  (xunxin)
* added the CompiledAction handler (currently for GET http commands only). (xunxin)
* now _callback and _var work with non-JSON data format as well.
* Model API
** use quasiquotes for function delete_records. (xunxin)
** added support for the Pg hstore column type.
** now we finally support _op = contains|le|ge|... correctly.
** now we support model columns of array types.
** started to stub out the prophet handler and model code (jesse)
* Version API
** fixed the revision inspection bug.
* Action API
** fixed the bug involved in params with unicode characters.
** fixed the bug involved in action call with param values of 0.
** added $_ROLE and $_ACCOUNT support.
** implemented HTTP forwarding to any sites in an action.
** added support for the frontend.allow_forwarding config option.
* View API
** added $_ROLE and $_ACCOUNT support.
* OpenResty.pm - enhances the DB re-connection stuff.
* Dispatcher - now we try to re-init at every request even if we have InitFatal.
* Captcha handler - now we use _lang rather than lang for languague indication.

* perl minisql compiler
** added support for the Pg array indexing syntax.
** now we allow proc() w/o arguments.

* haskell restyscript
** added support for subquery in from clause.
** added support for array indexing.
** added support for operators is and is not.
** added support for builtin variables.

* demo/Admin
** implemented the HTTP Console and Action Console panels.
** implemented incremental bulk insertion for rows.
** now the "Delete all rows" link can remove rows selectively
** enlarged the input boxes in the model creation form.
** added the "Go to end" link to the model row panel.
** sort the row fields by field name in both row display and row creation views.
** removed yahoo logo from the UI.
** added operator option list and other functionalities to the model row search panel.
** added support for exporting as CSV format and stream JSON format.
** added "prohibiting" to the columns for role spec.
** removed the "Click to edit" placeholder since it has been problematic.
** marked column labels to create model row panel.
* updated demo/BlogAdmin.
* demo/RestyCheck - now checks captcha images as well.
* demo/Blog
** now we sort comment list for a full post.
** fixed the ACL rule leaks in init.pl.
* demo/Springbot - now ignores fancybot and any other bots from now on.

* openresty.js
** worked around the cross-site POST issue in Opera.
** added support for urls *with* "?" in it


demo/Blog/init.pl
#!/usr/bin/env perl

use strict;
use warnings;

use lib '../../lib';
use utf8;
use JSON::Syck;
use YAML 'Dump';
use WWW::OpenAPI::Simple;
use Date::Manip;
use Getopt::Std;

my %opts;
getopts('u:s:p:h', \%opts);
if ($opts{h}) {
    die "Usage: $0 -u <user> -p <password> -s <openapi_server> [big|small] [<json_file>]\n";
}
my $user = $opts{u} or
    die "No OpenAPI account name specified via option -u\n";
my $password = $opts{p} or
    die "No OpenAPI account's Admin password specified via option -p\n";
my $server = $opts{s} || 'http://ced02.search.cnb.yahoo.com';

my $cmd = shift || 'small';
if ($cmd ne 'small' and $cmd ne 'big') {
    die "Unknown command: $cmd\n";
}

my $openapi = WWW::OpenAPI::Simple->new( { server => $server } );
$openapi->login($user, $password);
$openapi->delete("/=/model");
$openapi->delete("/=/role/Public/~/~");
$openapi->delete("/=/view");

$openapi->post({
    description => "Blog post",
    columns => [
        { name => 'title', label => 'Post title' },
        { name => 'content', label => 'Post content' },
        { name => 'author', label => 'Post author' },
        { name => 'created', default => ['now()'], type => 'timestamp(0) with time zone', label => 'Post creation time' },
        { name => 'comments', label => 'Number of comments', default => 0 },
    ],
}, '/=/model/Post');

$openapi->post({
    description => "Blog comment",
    columns => [
        { name => 'sender', label => 'Comment sender' },
        { name => 'email', label => 'Sender email address' },
        { name => 'url', label => 'Sender homepage URL' },
        { name => 'body', label => 'Comment body' },
        { name => 'created', default => ['now()'], type => 'timestamp(0) with time zone', label => 'Comment creation time' },
        { name => 'post', label => 'target post', type => 'integer' },
    ],
}, '/=/model/Comment');

print Dump($openapi->get('/=/model')), "\n";
#print Dump($openapi->get('/=/model/Post')), "\n";
#print Dump($openapi->get('/=/model/Comment')), "\n";
$openapi->post({
    definition => <<'_EOC_',
        select id, title, date_part('day', created) as day
        from Post
        where date_part('year', created) = $year and
              date_part('month', created) = $month
        order by created asc
_EOC_
}, '/=/view/PostsByMonth');

$openapi->post({
    definition => <<'_EOC_',
        select Comment.id as id, post, sender, title
        from Post, Comment
        where post = Post.id
        order by Comment.id desc
        offset $offset | 0
        limit $limit | 10
_EOC_
}, '/=/view/RecentComments');

$openapi->post({
    definition => <<'_EOC_',
        select id, title
        from Post
        order by id desc
        offset $offset | 0
        limit $limit | 10
_EOC_
}, '/=/view/RecentPosts');

$openapi->post({
    definition => <<'_EOC_',
        (select id, title
        from Post
        where id < $current
        order by id desc
        limit 1)
      union
        (select id, title
        from Post
        where id > $current
        order by id asc
        limit 1)
_EOC_
}, '/=/view/PrevNextPost');

$openapi->post({
    definition => <<'_EOC_',
        select count(*)
        from $model
_EOC_
}, '/=/view/RowCount');

$openapi->post([
    { method => "GET", url => '/=/model/Post/~/~' },
    { method => "GET", url => '/=/model/Comment/~/~' },

    { method => "GET", url => '/=/view/RecentComments/~/~' },
    { method => "GET", url => '/=/view/RecentPosts/~/~' },
    { method => "GET", url => '/=/view/PrevNextPost/~/~' },
    { method => "GET", url => '/=/view/PostsByMonth/~/~' },
    { method => "GET", url => '/=/view/RowCount/~/~' },

    { method => "POST", url => '/=/model/Comment/~/~' },
    { method => "PUT", url => '/=/model/Post/id/~' },
], '/=/role/Public/~/~');

################################################

if ($cmd eq 'small') {
    my ($title, $created, $buffer);
    while (<DATA>) {
        if (!$title) {
            next if /^\s*$/;
            $title = $_;
            chop $title;
        } elsif (!$created) {
            next if /^\s*$/;
            $created = $_;
            chop $created;
        } elsif (m{^////////+$}) {
            warn $title, "\n";
            $openapi->post({
                author => '章亦春',
                title => $title,
                content => $buffer,
                $created eq 'undef' ? () : (created => $created),
            }, '/=/model/Post/~/~');
            undef $title;
            undef $buffer;
            undef $created;
        } else {
            $buffer .= $_;
        }
    }
    if ($title && $buffer) {
        $openapi->post({
            author => '章亦春',
            title => $title,
            content => $buffer,
        }, '/=/model/Post/~/~');
    }

    #print Dump($openapi->get('/=/model/Post/~/~')), "\n";

    for my $i (1..5) {
        warn "Comment $i\n";
        $openapi->post([
            { sender => 'bot', body => qq{This is a comment <b>\t<a href="">&nbsp;</a>\t</b>$i\n} x 20, post => 2 },
        ], '/=/model/Comment/~/~');
        #sleep(0.8);
    }

    $openapi->post([
        { sender => 'laser', body => 'super cool!', url => 'www.pgsqldb.org', post => 4 },
        { sender => 'ting', body => '呵呵。。。', url => 'http://agentzh.org', post => 4 },
        { sender => 'clover', body => "yay!\nso great!", post => 3 },
    ], '/=/model/Comment/~/~');

    $openapi->put({ comments => 2 }, '/=/model/Post/id/4');
    $openapi->put({ comments => 1 }, '/=/model/Post/id/3');
    $openapi->put({ comments => 5 }, '/=/model/Post/id/2');
} else {
    my $infile = shift @ARGV || 'script/agentzh-live.json';
    open my $in, $infile or die "Can't open $infile for reading: $!\n";
    my $json = do { local $/; <$in> };
    my $data = JSON::Syck::Load($json);
    for my $entry (reverse @$data) {
        my $title = $entry->{title};
        my $body = $entry->{body};
        my $date = $entry->{date};
        warn $title, "\n";
        if ($body =~ /.{1,55}c\.gif.{3,150}/) {
            #warn $&, "\n";
        }
        $openapi->post({
            author => '章亦春',
            title => $title,
            content => $body,
            $date eq 'undef' ? () : (created => $date),
        }, '/=/model/Post/~/~');
    }
    $openapi->post([
        { sender => 'lobatt', body => "过来回拜：）\n关于你负责测试这个事情，我是在YAPC上获得的信息...不准确么？我改我改...", url => 'http://www.perlfect.org/', post => 62, created => date('November 25 8:47 PM') },
        { sender => 'agentzh', body => '呵呵，谢谢你的留言 :)', url => 'http://blog.agentzh.org', post => 62, created => date('November 27 8:43 PM') },
        { sender => 'xinglan', body => '非常好的一篇文章。期待下次崩溃 ：）', url => 'http://i.cn.yahoo.com/shi_xinglan', post => 65, created => '2007-12-03 11:09:00+08' },
    ], '/=/model/Comment/~/~');
    $openapi->put({ comments => 1 }, '/=/model/Post/id/65');
    $openapi->put({ comments => 2 }, '/=/model/Post/id/62');
}

sub date {
    my $date = shift;
    $date  =~ s/([A-Za-z]+ \d{1,2}) (\d+:\d+)/$1 2007 $2/;
    $date = ParseDate($date);
    my $str = UnixDate($date,'%Y-%m-%d %H:%M:%S+0800');
    print "$str\n";
    $str;
}

__DATA__

SearchAll 0.3.2 发布！
2007-08-09 13:05
经过 Yahoo! China EEEE hacking 小组近一个月的努力， <a href="http://searchall.agentzh.cn/">SearchAll</a>  0.3.2 版终于发布到了  <a href="http://addons.mozilla.org/">AMO 官方网站</a> ： </p><p><a href="https://addons.mozilla.org/zh-TW/firefox/addon/5712">
https://addons.mozilla.org/zh-TW/firefox/addon/5712</a> </p><p>如果您已经安装了 SearchAll 的话，您的 Firefox 在启动时会自动检查更新，并按用户选择进行升级。 </p><p>和上一个 AMO 公开发布版本 0.1.8 相比，SearchAll 积累了许多重大改进。其中的亮点包括： </p><ul><li> <p>增加了对结果页中的链接进行本地测试的功能。SearchAll 现在会自动在``规格化视图''中用可爱的小图标标记出坏链，好链，和慢链。由于使用 Ajax HEAD 请求，所以测试的速度很快。我们 team 的 
<a href="http://blog.jianingy.com/">杨家宁</a> 当初做 <a href="http://www.yisou.com/">"易搜"</a> 的时候就想到实现这样的本地链接测试的功能，但受到 AJAX 跨域请求的限制而未能实现。谢谢杨家宁提的这个 feature! </p></li><li> <p>增加了 <a href="http://agentzh.org/misc/rightclick.swf">

"划搜"功能</a> 。用户可以在任意的网页中选定文本，然后右击，在弹出的快捷菜单中点击"我搜去"（或者SearchAll)，进行搜索。感谢李晓栓的提议！ </p></li><li> <p>增
加了对简体中文(zh-CN)，繁体中文(zh-TW), 法语(fr)，和西班牙语(es-ES) 的国际化支持。如果用户使用的是简体中文版的
Firefox，那么她看到的 SearchAll 界面也将是简体中文版的。其他语言支持依此类推 :) 感谢西班牙的 Gregorio
Villarrubia 的西班牙文翻译。对于非中文用户而言，第一次使用 SearchAll 时默认的三家搜索引擎是 <a href="http://search.yahoo.com/">yahoo.com</a> ,  <a href="http://www.google.com/en">google.com</a> , 和  <a href="http://www.ask.com/">ask.com</a> . 中文用户默认的原三家网站保持不变。
 </p></li><li> <p>为大部分默认列表中的搜索引擎实现了  <a href="http://developer.mozilla.org/en/docs/Creating_OpenSearch_plugins_for_Firefox">OpenSearch 快捷方式</a> ，从而使第一次查询以及切换搜索引擎时的用时减少了至少 50%. 感谢服务于  <a href="http://www.ironport.com/">IronPort 公司

 </a> 的  <a href="http://wanghui.org/">cnhackTNT</a>  的提议！ </p></li><li> <p>修复了多处内存泄漏问题。使用 SearchAll 的时候，Firefox 不会再占用越来越多的内存。界面响应也更迅速。感谢 Mozilla 官方网站的编辑  <a href="http://wiki.mozilla.org/User:Archaeopteryx">Archaeopteryx</a>
   的报告。 </p></li><li> <p>SearchAll 工具条上的搜索框现在支持快捷菜单方式下的"复制"，"剪切"，和"粘贴"操作。感谢尚尔迪的报告和修复。 </p></li><li> <p>添加了下列新的默认搜索引擎： </p><ul><li> <p><a href="http://www.a9.com/">http://www.a9.com</a> </p></li><li> <p><a href="http://www.answers.com/">

   http://www.answers.com</a> </p></li><li> <p><a href="http://search.ebay.com/">http://search.ebay.com</a> </p></li><li> <p><a href="http://www.flickr.com/">http://www.flickr.com</a> </p></li><li> <p><a href="http://www.youtube.com/">
   http://www.youtube.com</a> </p></li></ul> </li></ul> <p>我们仍然在考虑如何在 SearchAll 中展现 Omni Search 的问题。既要保持 SearchAll 在结果展示上的公平公正性，又要表现 Omni Search 丰富的内容，并不是一件容易的事情 ;) 欢迎大家多提宝贵建议 :) </p><p>我最近在看有关  <a href="http://en.wikipedia.org/wiki/Adobe_Flex">

   Adobe Flex</a>  方面的文档，一个基于 Flash 的 SearchAll 似乎更酷一些，因为它将能运行在 IE 和 Opera 这样的浏览器中，而不仅仅是 Firefox. 当然这仅仅是一个很模糊的想法，不知是否可行。有兴趣的朋友可以与我们联系 ;) </p><p>SearchAll 是一个基于 MIT 协议的开源项目。源码仓库位于台湾的  <a href="http://rt.openfoundry.org/">OpenFoundry 服务器
   </a> 上：<a href="http://svn.openfoundry.org/searchall/trunk/">http://svn.openfoundry.org/searchall/trunk/</a>. 如果您乐意参与到项目开发中来，请立即与我们联系；我们很乐意像 Pugs 团队那样递送 Subversion 提交权限。 </p><p>感谢所有一直以来关注和使用 SearchAll 的朋友们；我们一如既往地欢迎各种新功能提议，bug 报告，和评论。
    </p><p>谢谢！ </p><p>章亦春 (agentzh) </p><p>附：根据陈敬亮的报告，来自  <a href="http://www.foxplus.org/">Foxplus</a>  的  <a href="https://addons.mozilla.org/en-US/firefox/addon/5362">Alexa Sparky 扩展</a> 会严重影响 SearchAll tab 的切换速度，因此对于安装了该版本的 Alexa Sparky 扩展的用户，请在使用 SearchAll 的时候暂时禁用 Alexa Sparky.

////////////////////////////

写了一篇东西到"雅虎搜索日志"
2006-12-31 12:27
写了一篇东西到我们 Yahoo 自己的"雅虎搜索日志"网站上：<br><br>&nbsp;&nbsp;&nbsp;&nbsp; <a href="http://ysearchblog.cn/2007/11/searchall.html">http://ysearchblog.cn/2007/11/searchall.html</a><br><br>感谢咱们 content team 的何远银同学提供的初稿。毕竟发起一些东西不是我的长项，而大刀阔斧地修改现成的东西却是，呵呵。
<br><br>另外还必须特别感谢一下我的编辑石杏岚小姐不厌其烦地反复修改这篇东西。在修改过 N 处之后，她终于说自己快崩溃了，呵呵。<br><br>这里顺便 spam 一下，SearchAll 的下一个版本将提供一个全新的视图，Mapping View：<br><br>&nbsp;&nbsp; <a href="http://agentzh.org/misc/mapview.png">http://agentzh.org/misc/mapview.png</a><br><br>欢迎大家试用 Subversion 里的版本：<br><br>&nbsp;&nbsp;&nbsp; <a href="http://svn.openfoundry.org/searchall/trunk/searchall.xpi">
http://svn.openfoundry.org/searchall/trunk/searchall.xpi</a><br><br>Enjoy!<br><br>-agentzh


/////////////////////////////

Yahoo! 4e team 贺岁语录
2008-02-09 02:05
"妈呀，又是测试啊？" -- leiyh<br><br>"话说……你加的功能 work 了！" -- carrie<br><br>"这事咋办呢？让我想想。。。" -- ting<br><br>"每日常坐电脑前，每逢春秋必感冒。锻炼永远计划中，感冒一直在行动。" -- jianingy<br><br>"谁说我不乖的？我很乖的。" -- ywayne
<br><br>"咦？是 exe 的。。。我来 hack 一下。。。" -- shangerdi<br><br>"哈哈！央视8套真好看,讲苍蝇飞行的原理 :D" -- laser<br><br>"锅得刚说了，铛铛铛，铛儿嘀儿嘀个儿铛！" -- arthas<br><br>"yay! it works! :D" -- agentzh<br><br>"春儿太猛了！整天写一坨一坨的没用的东西……" -- luoyi

////////////////////////////

作秀中...
2008-02-01 02:05
最近又写了一篇东西到"<a href="http://ysearchblog.cn/">雅虎搜索日志</a>"，题为"从SearchAll看搜索引擎DNA":<br><br>&nbsp;&nbsp;&nbsp; <a href="http://ysearchblog.cn/2008/01/searchalldna.html">http://ysearchblog.cn/2008/01/searchalldna.html</a><br><br>感谢杏岚的编辑工作 :)<br>
<br>我的下一篇东西可能题为"装在口袋里的网站"；我一直打算介绍一下我们的基于 OpenAPI 的纯客户端应用的开发技术。<br><br>我们的 M，yuting++，已经怪我"染上了作秀不良风气"了，哈哈。我看来是很难改正了，呵呵。tingting 一定要原谅我哦 ;)<br><br>-agentzh

/////////////////////////////

三叶草
undef
昨天在校园里漫步的时候，偶然间发现路旁的绿地上竟满是我最爱的三叶草。圆圆的叶子是那么娇嫩，那么雅致。我仿佛能觉出一缕淡淡的幽香……

////////////////////////////

怎样生活
undef
每天都在奔跑————不好，这不是我想要的生活。<br><br>每天都在学习————不好，这也不是我想要的生活。<br><br>每天都在工作————不好，这还不是我想要的生活。<br><br>我自己究竟想要什么样的生活呢？我不知道。<br><br>嗯，或许是那种比较清闲的生活。繁忙让我厌烦，压力让我不悦。<br><br>让别人拉着我向前奔跑？天哪，那有什么趣味？<br><br>让自己逼着我向前奔跑？嗨，奔跑又是为了什么？<br><br>按照“习俗”追求的许多东西，想来根本没有意义。归根究底，仍是放牛娃的逻辑：<br>学习，成功，娶妻，生子，学习，成功，娶妻，生子……<br><br>无聊之极。<br><br>极少去思考自己的思考。平日里所谓的思考，想来更多的是一种习惯，一种迷惑。<br><br>恶念在心中生起来却浑然不知，甚至乐此不彼。可悲，可叹，可恶！<br><br>终日只知愤世嫉俗，殊不知最俗最坏的竟是自己。<br><br>妍媸自彼，与我何干！<br><br>其实只愿在大部分的时间里，能够真真正正地静下心来，读懂我的灵魂，读懂我的心……<br><br>这个世界是虚幻的，我的心也是虚幻的。无奈。


